Unwrap row dereferences

martint opened the pull request:
Transforms expressions of the form

    CAST(ROW(x) AS ROW(f0 bigint)).f0

to

    CAST(x AS bigint)

kasiafi reviewed (commented):
Some high-level thoughts
1. I don't think it is correct to take a shortcut on row dereference. The whole base expression should be evaluated before the searched field is accessed. That was raised when the `ExpressionInterpreter` was being refactored recently. Otherwise it would be there already :) It is OK to unwrap dereference on a row that was synthetically created, but this is a general-purpose rewrite, so it applies to every `DereferenceExpression`.
2. Referring to the offline discussion, I think that it would be helpful to use the `SubscriptExpression` instead of `DereferenceExpression` for row dereferences. It would spare searching the field by name, which currently happens in several places. However, this is not the subject of this PR.
3. Could you consider extracting the "push cast into row" logic? It is not in conflict with (1), and could be useful with Values optimization.

martint commented:
I don't think (3) is as effective as we'd want until we have (2), since the field names assigned by the CAST may be meaningful.

phd3 commented:
> The whole base expression should be evaluated before the searched field is accessed.

@kasiafi can you please elaborate on why this is necessary or point to previous discussion? One reason I can think of is may be we should fail the query if query has issues computing any other fields.

If we're able to do this optimization of short-circuiting the row+dereference, it has a great practical usecase for views. People often create views on top of one or more tables combining fields into ROWs for more structured representation, but analytical queries don't reference all of them.

martint commented:
@phd3, @kasiafi and I had a related discussion a while ago around how ExpressionInterpreter works. But after thinking about it further, my conclusion is that we want row dereference to behave in the same way as column references. In particular, the following queries should behave similarly:

```
SELECT x
FROM (
    SELECT a AS x, 1/0
    FROM t
)
```

```
SELECT x
FROM (
    SELECT ROW(a, 1/0)[1] x
    FROM t
)
```

```
SELECT x[1]
FROM (
    SELECT ROW(a, 1/0) x
    FROM t
)
```

Technically, column pruning (or even predicate pushdown) may not conform to SQL spec semantics, which describes a strict evaluation order. But any implementation that follows that faithfully would be unusably slow in general.

martint committed:
Unwrap row dereferences

Transforms expressions of the form

ROW(x, y)[1] => x

CAST(ROW(x, y) AS row(f1 type1, f2 type2))[1] => CAST(x AS type1)

martint committed:
Push cast into row constructor

Transforms expressions of the form

    CAST(ROW(x, y, z) AS row(t1, t2, t3))

to

    ROW(CAST(x AS t1), CAST(y AS t2), CAST(y AS t3))

This can unlock other optimizations such as row dereference pushdown and pruning.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapRowDereference.java:
```diff
@@ -0,0 +1,80 @@
+                Expression item = row.getItems().get(index - 1);
+
+                if (coercion.isPresent()) {
+                    item = new Cast(item, coercion.get());
```
What about the `safe` and `typeOnly` parameters of Cast?

martint replied:
Ah, yes, it needs to handle `safe`, which is used for try_cast.

Let me think about `typeOnly`. It doesn't really matter (in fact, we should probably remove that concept from the engine altogether). It's more complicated to get that in here because we may need to re-analyze the expression types and check whether the cast is type-only. On the other hand, we may be able to infer type-only-ness based on the original cast, though, so let me look into it.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapRowDereference.java:
```diff
@@ -0,0 +1,80 @@
+            if (base instanceof Cast) {
+                Cast cast = (Cast) base;
+                if (cast.getType() instanceof RowDataType) {
+                    int index = (int) ((LongLiteral) node.getIndex()).getValue();
```
could you evaluate it once before the if?

martint replied:
No, because it's only valid to do so when it's subscript over a row type. Otherwise, node.getIndex() could be an arbitrary expression (i.e., for array access)

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestUnwrapRowDereference.java:
```diff
@@ -0,0 +1,54 @@
+        tester().assertThat(new UnwrapRowDereference().projectExpressionRewrite())
+                .on(p -> p.project(
+                        Assignments.builder()
+                                .put(p.symbol("output", BOOLEAN), expression(original))
```
Why `BOOLEAN`?

martint replied:
That's left-over from something I had before. I'll fix it.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestUnwrapRowDereference.java:
```diff
@@ -0,0 +1,54 @@
+        test("ROW(ROW(1, 2), 3)[1][2]", "2");
+
+        test("CAST(ROW(1, 2) AS row(a bigint, b bigint))[1]", "CAST(1 AS bigint)");
+        test("CAST(ROW(1, 2) AS row(bigint, bigint))[1]", "CAST(1 AS bigint)");
```
Could you add a test with a few levels of mixed Casts and Subscripts?

kasiafi reviewed (commented):
The two optimizations only work if there is a Row at the end of the chain.
It makes me think of a different approach.
We could push Subscript through Casts. If it reaches a nested Row, a field is extracted. If there is not a Row, but some other RowType expression, the Subscript remains on the innermost level. 
The result for nested Row would be the same, and we would get early pruning for other RowType expressions (plus avoid potential cast errors).
However
- implementing it would require some trick with context
- we still want "push cast into row" whenever possible, so there would be some duplication.
What do you think?

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapRowSubscript.java:
```diff
@@ -0,0 +1,111 @@
+            Expression base = treeRewriter.rewrite(node.getBase(), context);
+
+            Deque<Coercion> coercions = new ArrayDeque<>();
+            while (base instanceof Cast) {
+                Cast cast = (Cast) base;
+                if (cast.getType() instanceof RowDataType) {
+                    int index = (int) ((LongLiteral) node.getIndex()).getValue();
+                    base = cast.getExpression();
+                    RowDataType rowType = (RowDataType) cast.getType();
+                    coercions.push(new Coercion(rowType.getFields().get(index - 1).getType(), cast.isTypeOnly(), cast.isSafe()));
+                }
+            }
```
won't it loop if the Cast is not into `RowDataType`? E.g. in case of array subscript?

martint replied:
Yes, that's why the tests are failing :)

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapRowSubscript.java:
```diff
@@ -0,0 +1,111 @@
+                return item;
+            }
+
+            return treeRewriter.defaultRewrite(node, context);
```
if(!(base instanceof Row) && !coercions.isEmpty()) --> you can apply first the subscript, and then coercions.

martint replied:
That only works in we can prove `base` if of type `row(...)`. It could be done, but it complicates things further. I'll leave it as future improvement (will add a TODO)

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushCastIntoRow.java:
```diff
@@ -0,0 +1,55 @@
+        tester().assertThat(new PushCastIntoRow().projectExpressionRewrite())
+                .on(p -> p.project(
+                        Assignments.builder()
+                                .put(p.symbol("output", BOOLEAN), expression(original))
```
remove `BOOLEAN`

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushCastIntoRow.java:
```diff
@@ -0,0 +1,55 @@
+        test("CAST(CAST(ROW(1, 'a') AS row(smallint, varchar)) AS ROW(bigint, varchar))", "ROW(CAST(CAST(1 AS smallint) AS bigint), CAST(CAST('a' AS varchar) AS varchar))");
+
+        // named fields in top-level cast preserved
+        test("CAST(CAST(ROW(1, 'a') AS row(smallint, varchar)) AS ROW(x bigint, varchar))", "CAST(ROW(CAST(1 AS smallint), CAST('a' AS varchar)) AS ROW(x bigint, varchar))");
```
Add a test with named fields on intermediate level.
Add a test where the rewritten expression is nested (e.g. inside subscript).

It would be good to have a test which shows how the two rules cooperate.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapRowSubscript.java:
```diff
@@ -0,0 +1,121 @@
+                int index = (int) ((LongLiteral) node.getIndex()).getValue();
+                RowDataType rowType = (RowDataType) cast.getType();
+                DataType type = rowType.getFields().get(index - 1).getType();
+                if (!(type instanceof GenericDataType) || !((GenericDataType) type).getName().getCanonicalValue().equals("UNKNOWN")) {
```
Why are the casts into UNKNOWN skipped? Is it a fix or an optimization?
Should `PushCastIntoRow` do it too?
Also, I think we resolve the type name case insensitive. So, when filtering by the canonical name, some occurrences might be missed.

martint replied:
`CAST(x AS unknown)` is an invalid statement (`unknown` is not a real type. It's a synthetic type to represent "any type"). Without this check, when type of the row is inferred as `row(unknown)`, this optimizer will try to unwrap it and insert an invalid cast to unknown.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushCastIntoRow.java:
```diff
@@ -0,0 +1,81 @@
+ * Transforms expressions of the form
+ *
+ * <pre>
+ *     CAST(CAST(ROW(x, y) AS row(f1 type1, f2 type2) AS row(g1 type3, g2 type4)
```
not quite here :)

martint replied:
Oh, right. Let me fix that.
